File: src/main/java/com/example/geomath/Vector2D.java
package com.example.geomath;

import java.util.Objects;

public final class Vector2D {
    private final double x;
    private final double y;

    public Vector2D(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double x() { return x; }
    public double y() { return y; }

    public Vector2D add(Vector2D other) {
        return new Vector2D(this.x + other.x, this.y + other.y);
    }

    public Vector2D subtract(Vector2D other) {
        return new Vector2D(this.x - other.x, this.y - other.y);
    }

    public Vector2D scale(double k) {
        return new Vector2D(this.x * k, this.y * k);
    }

    public double dot(Vector2D other) {
        return this.x * other.x + this.y * other.y;
    }

    public double cross(Vector2D other) {
        return this.x * other.y - this.y * other.x;
    }

    public double norm() {
        return Math.hypot(this.x, this.y);
    }

    /**
     * Rotate this vector counter-clockwise by the given angle.
     * @param angle angle value
     * @param degrees if true, angle is in degrees; otherwise radians
     *
     * INTENTIONAL BUG (for the demo):
     * - When degrees==true, this currently treats the input as radians.
     */
    public Vector2D rotate(double angle, boolean degrees) {
        double theta = angle; // BUG: should convert when degrees==true
        // if (degrees) {
        //     theta = Math.toRadians(angle);
        // }
        double c = Math.cos(theta);
        double s = Math.sin(theta);
        return new Vector2D(this.x * c - this.y * s, this.x * s + this.y * c);
    }

    public boolean almostEquals(Vector2D other, double tol) {
        return Math.abs(this.x - other.x) <= tol && Math.abs(this.y - other.y) <= tol;
    }

    @Override
    public String toString() {
        return "Vector2D(" + x + ", " + y + ")";
    }

    // Deliberately do not override equals/hashCode for exact FP equality,
    // tests use almostEquals instead.
    @Override
    public boolean equals(Object o) {
        return super.equals(o);
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }
}

File: geomath/vector2d.py
"""2D vector helpers for the geomath demo project.

This module intentionally mirrors the original Java implementation.
It includes the same deliberate bug in :meth:`Vector2D.rotate` where the
``degrees`` flag is ignored.
"""

from __future__ import annotations

from dataclasses import dataclass
import math


@dataclass(frozen=True)
class Vector2D:
    """Immutable two dimensional vector."""

    x: float
    y: float

    def add(self, other: "Vector2D") -> "Vector2D":
        """Return the vector sum of ``self`` and ``other``."""

        return Vector2D(self.x + other.x, self.y + other.y)

    def subtract(self, other: "Vector2D") -> "Vector2D":
        """Return the vector difference ``self - other``."""

        return Vector2D(self.x - other.x, self.y - other.y)

    def scale(self, k: float) -> "Vector2D":
        """Return the vector scaled by ``k``."""

        return Vector2D(self.x * k, self.y * k)

    def dot(self, other: "Vector2D") -> float:
        """Return the dot product of the vectors."""

        return self.x * other.x + self.y * other.y

    def cross(self, other: "Vector2D") -> float:
        """Return the 2D cross product magnitude (signed area)."""

        return self.x * other.y - self.y * other.x

    def norm(self) -> float:
        """Return the Euclidean magnitude of the vector."""

        return math.hypot(self.x, self.y)

    def rotate(self, angle: float, degrees: bool) -> "Vector2D":
        """Rotate the vector counter-clockwise by ``angle``.

        The ``degrees`` flag is supposed to indicate the angle unit. This
        implementation intentionally ignores the flag to preserve the bug
        present in the original Java version.
        """

        theta = angle  # BUG: should convert when degrees is True
        c = math.cos(theta)
        s = math.sin(theta)
        return Vector2D(self.x * c - self.y * s, self.x * s + self.y * c)

    def almost_equals(self, other: "Vector2D", tol: float) -> bool:
        """Return ``True`` when the coordinates differ by at most ``tol``."""

        return abs(self.x - other.x) <= tol and abs(self.y - other.y) <= tol
